Refactor remaining local exceptions in C#

bernardnormier opened the pull request:
This PR renames the following exceptions:
CloseConnectionException -> ConnectionClosedByPeerException
ConnectionClosedManuallyException -> ConnectionClosedLocallyException (and drops Graceful)
Timeout -> ConnectionTimeoutException

The new InvocationTimeout exception is now an "abstract" System.Exception with two derived classes: InvocationTimeoutException and InvocationCanceledException.

ConnectTimeoutException now derives from ConnectFailedException.

bernardnormier commented on csharp/test/Ice/metrics/AllTests.cs:
```diff
@@ -627,9 +627,8 @@ public static IMetricsPrx allTests(TestHelper helper, CommunicatorObserver obsv)
                 Thread.Sleep(10);
             }
-            test(cm1.Failures == 2 && sm1.Failures >= 2);
+            test(cm1.Failures == 1 && sm1.Failures >= 2);
```
It's unclear why the exception renaming/refactoring changed this number.

pepone replied:
Seems like the invocation was causing a retry before, but not anymore.

pepone commented on csharp/test/Ice/metrics/AllTests.cs:
```diff
@@ -612,7 +612,7 @@ public static IMetricsPrx allTests(TestHelper helper, CommunicatorObserver obsv)
                 metrics.Clone(connectionTimeout: 500).opByteS(new byte[10000000]);
```
not clear why we do this clone if we are setting the same timeout, isn't this returning the same proxy?

pepone replied:
nvm the previous call didn't assign the proxy

bernardnormier commented:
Add new exception class ConnectionIdleTimeout (as suggesed by Benoit), renamed InvocationTimeoutException -> System.TimeoutException, InvocationCanceledException -> System.OperationCanceledException, fixed tests.

bentoi commented on csharp/src/Ice/Connection.cs:
```diff
@@ -677,7 +677,7 @@ public void AsyncRequestCanceled(OutgoingAsyncBase outAsync, System.Exception ex
                     }
 
-                    if (ex is ConnectionTimeoutException)
+                    if (ex is ConnectionIdleException)
```
This isn't correct. The ConnectionTimeoutException check here was to support `invocationTimeout = -2` which is supposed to close the connection on invocation timeout. You changed the invocation timeout cancellation to now use TimeoutException for both `invocationTimeout = -2` and `!= -2` so there's no way to distinguish anymore between these 2 cases. We should either remove support for invocationTimeout = -2 and this `if()` block or allow the code to distinguish between the 2 reasons for invocation timeouts.

bernardnormier replied:
Fixed by removing `invocationTimeout = -2`. It's no longer a valid value.

bentoi commented on csharp/src/Ice/Connection.cs:
```diff
@@ -710,7 +710,7 @@ public void AsyncRequestCanceled(OutgoingAsyncBase outAsync, System.Exception ex
                         if (kvp.Value == outAsync)
                         {
-                            if (ex is ConnectionTimeoutException)
+                            if (ex is ConnectionIdleException)
```
Same issue there, ex is never ConnectionIdleException here.

bernardnormier replied:
Fixed (removed)

bentoi commented on csharp/src/Ice/OutgoingAsync.cs:
```diff
@@ -593,17 +595,9 @@ protected override bool ResponseImpl(bool userThread, bool ok, bool invoke)
-                Cancel(new Ice.InvocationTimeoutException());
-            }
-        }
+            => Cancel(new TimeoutException($"invocation on remote Ice object `{Proxy.Identity}' timed out"));
```
Using the same exception effectively removes the support for `invocationTimeout = -2`.

bentoi commented on csharp/test/Ice/retry/AllTests.cs:
```diff
@@ -240,26 +241,26 @@ static public Test.IRetryPrx
-                    // ensure retries with the old connection timeout semantics work.
+                    // The old invocationTimeout = -2 behaves like any timeout now, with no retry
+                    // TODO: should we remove this entirely? can the connection establishment timeout and trigger a
+                    // retry?
```
I don't see the point of introducing new semantics for invocationTimeout = -2, I think we should just remove support for it.

bernardnormier replied:
Done in latest commit
